src/lib/backend_fapi.c

```diff
@@ -10,6 +10,96 @@
 FAPI_CONTEXT *fctx = NULL;
 unsigned maxobjectid = 0;
 
+#ifndef FAPI_3_0
+/* The following piece of code is here to enable a preview of the FAPI backend on systems with a
+   TSS < 3.0, and retain the current "FAPI_PREVIEW" feature. */
+
+#include <tss2/tss2_esys.h>
+
+#define FAPI_ESYSBLOB_CONTEXTLOAD 1
+#define FAPI_ESYSBLOB_DESERIALIZE 2
+TSS2_RC Fapi_GetEsysBlob(FAPI_CONTEXT *fctx, const char *path, uint8_t *type, uint8_t **data, size_t *length) {
+
+    TSS2_RC rc;
+    TSS2_TCTI_CONTEXT *tcti;
+    ESYS_CONTEXT *esys;
+    ESYS_TR esystr;
+
+    if (!strstr(path, "/HS/SRK")) {
+        return TSS2_FAPI_RC_PATH_NOT_FOUND;
+    }
+    rc = Fapi_GetTcti(fctx, &tcti);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOGE("Fapi_GetTcti");
+        return rc;
+    }
+    rc = Esys_Initialize(&esys, tcti, NULL);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOGE("Esys_Initialize");
+        return rc;
+    }
+    rc = Esys_TR_FromTPMPublic(esys, 0x81000001, ESYS_TR_NONE, ESYS_TR_NONE, ESYS_TR_NONE, &esystr);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOGE("Esys_TR_FromTPMPublic");
+        return rc;
+    }
+    rc = Esys_TR_Serialize(esys, esystr, data, length);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOGE("Esys_TR_Serialize");
+        return rc;
+    }
+    Esys_Finalize(&esys);
+
+    *type = FAPI_ESYSBLOB_DESERIALIZE;
+    return TSS2_RC_SUCCESS;
+}
+#endif /* FAPI_3_0 */
+
+static CK_RV get_key(FAPI_CONTEXT *fctx, tpm_ctx *tctx, const char *path, uint32_t *esysHandle, uint32_t *tpmHandle) {
+
+    bool ret;
+    TSS2_RC rc;
+    uint8_t type;
+    uint8_t *data;
+    size_t length;
+
+    rc = Fapi_GetEsysBlob(fctx, path, &type, &data, &length);
+    if (rc != TSS2_RC_SUCCESS) {
+        LOGE("Cannot get Esys blob for key %s", path);
+        return CKR_GENERAL_ERROR;
+    }
+
+    twist twistdata = twistbin_new(data, length);
+    Fapi_Free(data);
```

Comment:
Will this leak things like data from Fapi_GetEsysBlob?

Reply:
I don't understand...

Reply:
> I don't understand...

I don't understand either, I am an idoit